src/lib/mina_generators/parties_generators.ml

```diff
@@ -1002,20 +1071,30 @@ let gen_fee_payer ?failure ?permissions_auth ~account_id ~ledger
 *)
 let max_other_parties = 2
 
-let gen_parties_from ?(failure = None)
-    ~(fee_payer_keypair : Signature_lib.Keypair.t)
+let gen_parties_from ?failure ~(fee_payer_keypair : Signature_lib.Keypair.t)
     ~(keymap :
        Signature_lib.Private_key.t Signature_lib.Public_key.Compressed.Map.t )
-    ~ledger ?protocol_state_view ?vk ?prover () =
+    ?account_state_tbl ~ledger ?protocol_state_view ?vk ?prover () =
   let open Quickcheck.Let_syntax in
   let fee_payer_pk =
     Signature_lib.Public_key.compress fee_payer_keypair.public_key
   in
   let fee_payer_account_id = Account_id.create fee_payer_pk Token_id.default in
-  let ledger_accounts = Ledger.accounts ledger in
+  let ledger_accounts = Ledger.to_list ledger in
+  (* table of public keys to accounts, updated when generating each party
+
+     a Map would be more principled, but threading that map through the code
+     adds complexity
+  *)
+  let account_state_tbl =
+    Option.value account_state_tbl ~default:(Account_id.Table.create ())
+  in
   (* make sure all ledger keys are in the keymap *)
-  Account_id.Set.iter ledger_accounts ~f:(fun acct_id ->
+  List.iter ledger_accounts ~f:(fun acct ->
+      let acct_id = Account.identifier acct in
       let pk = Account_id.public_key acct_id in
+      (*Initialize account states*)
+      Account_id.Table.update account_state_tbl acct_id ~f:(fun _ -> acct) ;
```

Comment:
Since the states in `account_state_tbl` is newer than the ones in ledger, I think we don't want to overwrite the ones in `account_state_tbl`:
```
      Account_id.Table.change account_state_tbl acct_id ~f:(function
        | None -> Some acct 
        | Some acct' -> Some acct') ;
```